candid_diff/src/candiff.rs

```diff
@@ -0,0 +1,298 @@
+                },
+                CliFormat::Raw => {
+                    // how is it encoded as a string?
+                    // base64?
+                    // always take binary from stdin?
+                    // something else?
+                    unimplemented!()
+                }
+            }
+        }
+        CliCommand::DiffValue{ format, input1, input2 } => {
+            match format {
+                CliFormat::Text => {
+                    match (Value::from_str(&input1),
+                           Value::from_str(&input2)) {
+                        (Err(e1), Err(e2)) => { error!("Both values failed to parse:\nFirst: {}\nSecond: {}", e1, e2) },
+                        (Err(e1), _) => { error!("First value failed to parse (only): {}\n", e1) },
+                        (_, Err(e2)) => { error!("Second value failed to parse (only): {}\n", e2) },
+                        (Ok(v1), Ok(v2)) => {
+                            trace!("value_1 = {:?}", v1);
+                            trace!("value_2 = {:?}", v2);
+                            let edit = candid_diff::value_diff(&v1, &v2, None);
+                            trace!("value_diff = {:?}", edit.0);
+                            if candid_diff::value_edit_is_skip(&edit) {
+                                debug!("equal values; no change.")
+                            } else {
+                                // to do -- more minimal textual output format for edits
+                                println!("{:?}", edit.0)
+                            }
+                        }
+                    }
+                }
+                CliFormat::Raw => {
+                    // how is it encoded as a string?
+                    // base64?
+                    // always take binary from stdin?
+                    // something else?
+                    unimplemented!()
+                }
+            }
+        }
+        CliCommand::Completions { shell: s } => {
+            // see also: https://clap.rs/effortless-auto-completion/
+            //
+            CliOpt::clap().gen_completions_to("candiff", s, &mut std::io::stdout());
+            info!("done")
+        }
+    }
+}
+
+/// Unit tests for command line interface (move elsewhere?)
+#[cfg(test)]
+mod cli {
+    use predicates::prelude::*;
+    use assert_cmd::Command;
+
+    #[test]
+    fn help() {
+        let mut cmd = Command::cargo_bin("candiff").unwrap();
+        cmd.arg("help");
+        cmd.assert().success();
+    }
+
+    #[test]
+    fn echo_number() {
+        let mut cmd = Command::cargo_bin("candiff").unwrap();
+        cmd.arg("echo").arg("1");
+        cmd.assert()
+            .stdout(predicate::eq(b"1\n" as &[u8]))
+            .success();
+    }
+
+    #[test]
+    fn echo_number_debug() {
+        let mut cmd = Command::cargo_bin("candiff").unwrap();
+        cmd.arg("echo").arg("1").arg("-d");
+        cmd.assert()
+            .stdout(predicate::eq(b"Number(\"1\")\n" as &[u8]))
```

Comment:
@chenyan-dfinity Here's an example of how the parser works now: I always get back a `Number`, and never a `Nat` or `Int`, etc.  Unfortunately, I don't think this is the behavior we want for the diff tool.  Thoughts?

Reply:
Yes, the current assumption is that the values will always have an associated candid file, and you need to call https://github.com/dfinity/candid/blob/master/candid/src/parser/value.rs#L176 to get the non-Number type. Specifically, the types listed here https://github.com/dfinity/candid/blob/master/candid/src/parser/value.rs#L23 won't be derived without a candid type.

Reply:
Cool, makes perfect sense.  Thanks for confirming.

Reply:
In general, you won't know the type if the parser only sees a number `1`, and you cannot enforce the input to the diff tool to annotate all number types (the parser don't support type annotations yet). For serialization, we always treat `Number` as `Int`, if it is unresolved for whatever reason.

Reply:
I've added some features and tests that permit the user to give the type on the command line.  Check out the tests in `candid.rs` to see use cases.

WDYT?

Reply:
Parsing as `IDLType` won't work for recursive types, you need a candid file for that. As a first step, it's actually not too bad to treat every number as `Number` without the need for type annotations.

Reply:
>  you need a candid file for that

Definitely.  I have left a comment in the code where I'd supply a non-empty type environment (from one or more `.did` inputs)